main/broadcasters.c

```diff
@@ -389,12 +388,78 @@ static broadcaster_ops_t mijia_temp_hum_ops = {
     .is_broadcaster = mijia_temp_hum_is_broadcaster,
     .metadata_get = mijia_temp_hum_metadata_get,
 };
+/* Beewi Smart Door */
+#define BEEWI_SMART_DOOR_COMPANY_ID     0x000D
+#define BEEWI_SMART_DOOR_SERVICE_ID     0x08 
+#define BEEWI_SMART_DOOR_DATA_TYPE_STAT 0x0C
+#define BEEWI_SMART_DOOR_DATA_TYPE_BATT 0x06
+
+
+typedef struct {
+    uint16_t company_id;
+    uint8_t  service_id;
+	uint8_t  type_stat;
+    uint8_t  status;
+	uint8_t  type_batt;
+	uint8_t  battery;
+} __attribute__((packed)) beewi_smart_door_t;
+
+static beewi_smart_door_t *beewi_smart_door_data_get(uint8_t *adv_data, uint8_t adv_data_len,
+    uint8_t *beewi_smart_door_len)
+{
+    uint8_t len;
+    uint8_t *data = esp_ble_resolve_adv_data(adv_data,
+        ESP_BLE_AD_MANUFACTURER_SPECIFIC_TYPE, &len);
+
+    if (beewi_smart_door_len)
+        *beewi_smart_door_len = len;
+
+    return (beewi_smart_door_t *)data;
+}
+
+static int beewi_smart_door_is_broadcaster(uint8_t *adv_data, size_t adv_data_len)
+{
+    uint8_t len;
+    beewi_smart_door_t *beewi_smart_door = beewi_smart_door_data_get(adv_data, adv_data_len, &len);
+
+    if (!beewi_smart_door || len != sizeof(beewi_smart_door_t))
+        return 0;
+
+    /* Technically, we should also prevent connecting to the device, as it stops broadcasting 
+       then. Or just once to get device name, and other static infos. Interesting data as opening status & battery are broadcasted. 
+	   Battery data is also available as characteristic, but not the essential data, the door status.
+	   Also it seems to lock the device ( todo: see wtf happening. )
+	   Needs to blacklist the mac in service config to make this working*/
+	   
+    return le16toh(beewi_smart_door->company_id) == BEEWI_SMART_DOOR_COMPANY_ID;
+}
+
+static void beewi_smart_door_metadata_get(uint8_t *adv_data, size_t adv_data_len,
+    int rssi, broadcaster_meta_data_cb_t cb, void *ctx)
+{
+    char s[6];
+    beewi_smart_door_t *beewi_smart_door = beewi_smart_door_data_get(adv_data, adv_data_len, NULL);
+    if ((beewi_smart_door->service_id == BEEWI_SMART_DOOR_SERVICE_ID) &&
```

Comment:
This validation should be done as a part of `beewi_smart_door_is_broadcaster()`, by the time this function is called you should already be sure the device type is a Beewi smart door sensor and just purse the relevant data.